### PR TITLE
Add GeniePlus Harvester

### PR DESCRIPTION
**Pull request by demiankatz**
TODO
- [x] Merge in work from #98 when completed

### CODE REVIEW

**Comment by demiankatz** [comment]
@EreMaijala, please consider this a draft for now, but I wanted to share my work in progress in case you have any feedback.

File: conf/datasources.ini.sample [hunk 1]
```diff
@@ -187,3 +187,33 @@ institution = SierraLib
 format = marc
 dedup = true
 building_mapping = sierra_locations.map
+
+; Sample configuration for GeniePlus REST API
+[GeniePlus]
+type = GeniePlus
+genieplusDatabase = "[api_database_name]"
```
Comment (EreMaijala) [inline conf/datasources.ini.sample]: This one as well as `genieplusTemplate` have a lowercase p, should be unified.

File: src/RecordManager/Base/Harvest/GeniePlus.php [hunk 2]
```diff
@@ -0,0 +1,682 @@
...
+            );
+        }
+        $this->database = $settings['genieplusDatabase'];
+        $this->template = $settings['genieplusTemplate'] ?? 'Catalog';
+        $this->oauthId = $settings['geniePlusOauthId'];
+        $this->username = $settings['geniePlusUsername'];
+        $this->password = $settings['geniePlusPassword'];
+        $this->batchSize = $settings['batchSize'] ?? 100;
+        $this->idField = $settings['geniePlusIdField'] ?? 'UniqRecNum';
+        $this->marcField = $settings['MarcRecord'] ?? 'MarcRecord';
```
Comment (EreMaijala) [inline src/RecordManager/Base/Harvest/GeniePlus.php]: Should this be `$settings['geniePlusMarcField']`?

File: src/RecordManager/Base/Harvest/GeniePlus.php [hunk 3]
```diff
@@ -0,0 +1,682 @@
...
+        $replacementCount = 0;
+        $rawXml = preg_replace(
+            '/[^\x{0009}\x{000a}\x{000d}\x{0020}-\x{D7FF}\x{E000}-\x{FFFD}]+/u',
+            '',
+            $this->lineBasedFormatter->convertLineBasedMarcToXml($marc),
+            -1,
+            $replacementCount
+        );
+        if ($replacementCount > 0) {
+            $this->infoMsg("Replaced $replacementCount bad chars in record $id");
```
Comment (EreMaijala) [inline src/RecordManager/Base/Harvest/GeniePlus.php]: infoMsg may be fine, but consider warningMsg/debugMsg as well depending on how frequent occurrence this is and how much bearing it has on the resulting records.

**Review by EreMaijala — CHANGES_REQUESTED** [review]

**Comment by EreMaijala** [comment]
@demiankatz Looks pretty good to me already. I glanced through the code and added a couple of observations.

File: src/RecordManager/Base/Harvest/GeniePlus.php (hunk 3 above)
Comment (demiankatz) [reply src/RecordManager/Base/Harvest/GeniePlus.php]: Yes, I think warning is better -- changed (along with merging in #98).

File: src/RecordManager/Base/Harvest/GeniePlus.php (hunk 2 above)
Comment (demiankatz) [reply src/RecordManager/Base/Harvest/GeniePlus.php]: Fixed!

File: conf/datasources.ini.sample (hunk 1 above)
Comment (demiankatz) [reply conf/datasources.ini.sample]: Fixed!

**Comment by demiankatz** [comment]
Thanks, @EreMaijala, I think we're in pretty good shape now -- I've addressed all of your feedback. Feel free to merge this if you think it's ready, or to leave it open if you want to give it a few days to be sure I don't make more changes. I think I've met all of my immediate needs, but it's possible I might come up with a tweak or two as I test further.

**Review by EreMaijala — APPROVED** [review]
I can't actually test this, but looks good to me!

**Pull request merged**